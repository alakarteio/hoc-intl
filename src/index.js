export { default } from './core'
export { addLocaleData } from './utils'
