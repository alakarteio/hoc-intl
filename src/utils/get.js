const get = (data, path) => path.split('.').reduce(
  (curr, sub) => curr && curr[sub],
  data,
)

export default (data, path) => (typeof path === 'string' ? get(data, path) : data)
