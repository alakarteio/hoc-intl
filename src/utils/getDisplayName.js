export default Component => `Intl(${
  Component.displayName
  || Component.name
  || (Component.constructor && Component.constructor.name)
  || 'Unknown'
})`
