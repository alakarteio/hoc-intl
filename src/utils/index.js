export { default as wrapDisplayName } from './wrapDisplayName'
export { default as getLocale } from './getLocale'
