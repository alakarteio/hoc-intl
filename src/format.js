// eslint-disable-next-line import/no-unresolved, import/extensions
import defaultLocaleData from 'k-intl/locale-data/en'
import { get } from 'lodash'
import IntlMessageFormat from 'intl-messageformat'
import { addLocaleData } from './utils'

const defaultMessages = {}
const browserLanguage = window.navigator.language || window.navigator.browserLanguage
/* add default locale on startup, import on src folder */
addLocaleData(defaultLocaleData)

export const formatter = (...args) => new IntlMessageFormat(...args)

const formatMessage = (params, message = '', ...args) => {
  if (params) {
    try {
      return formatter(message, ...args).format(params)
    } catch (e) {
      console.warn('/HOC k-intl/', e.message) // eslint-disable-line no-console
    }
  }
  return message
}

const getKey = (path) => {
  const splitPath = path.split('.')
  return splitPath[splitPath.length - 1]
}

const getSubPaths = (locale, paths, config) => {
  const subMessages = get(locale, paths)

  if (!subMessages || typeof subMessages === 'string') {
    return { [getKey(paths)]: paths }
  }

  return Object
    .keys(subMessages)
    .reduce(
      (acc, curr) => ({ ...acc, [curr]: `${config}.${curr}` }),
      defaultMessages,
    )
}

export default
(lang = browserLanguage, locale, customFormats) =>
  (config = {}, componentProps = []) => {
    // in case of locale is not initialized
    if (!locale) return {}

    let paths = config
    if (typeof paths === 'string') paths = getSubPaths(locale, paths, config)
    return Object
      .keys(paths)
      .reduce((acc, curr) => {
        const message = get(locale, paths[curr])
        return { ...acc, [curr]: formatMessage(componentProps[curr], message, lang, customFormats) }
      }, defaultMessages)
  }
