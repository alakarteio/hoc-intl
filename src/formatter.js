// eslint-disable-next-line import/no-unresolved, import/extensions
import IntlMessageFormat from 'intl-messageformat'
import { get, getSubPaths } from './utils'

const formatMessage = (message = '', ...args) => (values) => {
  if (values) {
    try {
      return new IntlMessageFormat(message, ...args).format(values)
    } catch (e) {
      return message
    }
  }
  return message
}

const browserLanguage = window.navigator.language || window.navigator.browserLanguage

export default (lang = browserLanguage, locale, formats) => (rootPath = '', values = {}) => {
  // in case of locale is not initialized
  if (!locale) return {}

  let paths = rootPath
  if (typeof paths === 'string') paths = getSubPaths(locale, paths, rootPath)
  return Object
    .keys(paths)
    .reduce((acc, curr) => {
      const message = get(locale, paths[curr]) || ''
      return { ...acc, [curr]: formatMessage(message, lang, formats)(values[curr]) }
    }, {})
}
