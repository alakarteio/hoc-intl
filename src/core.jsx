import React, { Component } from 'react'
import shallowEqual from 'fbjs/lib/shallowEqual'
import { wrapDisplayName, getLocale } from './utils'
import format from './format'

export default config => WrappedComponent => class extends Component {
  static displayName = wrapDisplayName(WrappedComponent, 'Intl')

  static contextTypes = {
    store: () => null, // this is to avoid importing prop-types
  }

  constructor(props, context) {
    super(props, context)
    this.labels = {}
    this.state = {
      injectedProps: {},
    }
  }

  componentWillMount() {
    this.unsubscribe = this.context.store.subscribe(() => { this.format() })
    this.format()
  }

  componentWillReceiveProps(nextProps) {
    this.format(nextProps)
  }

  componentWillUnmount() {
    this.unsubscribe()
  }

  format = (nextProps) => {
    /* take locale on `config.locale` reducer */
    const locale = getLocale(this.context)
    const labels = format(locale)(
      config,
      nextProps || this.props,
    )
    /* not change labels when the formated labels is identical */
    if (shallowEqual(this.state.injectedProps.labels, this.labels)) return
    this.setState(state => ({
      ...state,
      injectedProps: { labels },
    }))
  }

  render() {
    return (
      <WrappedComponent
        /* this is parent props */
        {...this.props}
        /* this is injected props from hoc */
        {...this.state.injectedProps}
      />
    )
  }
}
