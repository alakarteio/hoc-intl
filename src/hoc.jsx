import React, { Component } from 'react'
import shallowEqual from 'fbjs/lib/shallowEqual'
import { wrapDisplayName, getConfigs } from './utils'
import formatter from './formatter'

export default (path, getState) => WrappedComponent => class extends Component {
  static displayName = wrapDisplayName(WrappedComponent, 'Intl')

  static contextTypes = {
    store: () => null, // this is to avoid importing prop-types
  }

  constructor(props, context) {
    super(props, context)
    this.messages = {}
    this.state = {
      injectedProps: {},
    }
  }

  componentWillMount() {
    this.unsubscribe = this.context.store.subscribe(() => { this.inject() })
    this.inject()
  }

  componentWillReceiveProps(nextProps) {
    this.inject(nextProps)
  }

  componentWillUnmount() {
    this.unsubscribe()
  }

  inject = (nextProps) => {
    /* take lang, locale and formats on `store` reducer */
    const { lang, locale, formats } = getConfigs(getState)(this.context)

    /* path can be a function, in which case we pass props to it */
    let innerPath = path
    if (typeof path === 'function') innerPath = path(nextProps || this.props)

    /* format messages */
    const messages = formatter(lang, locale, formats)(
      innerPath,
      nextProps || this.props,
    )
    /* not change messages when the formated messages is identical */
    if (shallowEqual(this.state.injectedProps.messages, messages)) return
    this.setState(state => ({
      ...state,
      injectedProps: { messages },
    }))
  }

  render() {
    return (
      <WrappedComponent
        /* this is parent props */
        {...this.props}
        /* this is injected props from hoc */
        {...this.state.injectedProps}
      />
    )
  }
}
