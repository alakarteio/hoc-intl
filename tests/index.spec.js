/* eslint-env jest */
import format, { formatter } from '../src/format'
import { addLocaleData } from '../src/utils'
import localDataFR from '../locale-data/fr'
import en from './resources/en-GB'
import fr from './resources/fr-FR'

const customFormats = {
  number: {
    usd: { style: 'currency', currency: 'USD' },
  },
}

const keyValueResources = (lang, f) => ({
  language: formatter('', lang).resolvedOptions(),
  'simple.withoutParam': f({ test: 'simple.withoutParam' }),
  'simple.withOneParam': f({ test: 'simple.withOneParam' }, { test: { one: 'ONE' } }),
  'simple.withTwoParam': f({ test: 'simple.withTwoParam' }, { test: { one: 'ONE', two: 'TWO' } }),
  'plural.simple.noparam': f({ test: 'plural.simple' }),
  'plural.simple.zero': f({ test: 'plural.simple' }, { test: { param: 0 } }),
  'plural.simple.one': f({ test: 'plural.simple' }, { test: { param: 1 } }),
  'plural.simple.two': f({ test: 'plural.simple' }, { test: { param: 2 } }),
  'plural.simple.thousand': f({ test: 'plural.simple' }, { test: { param: 1000 } }),
  'number.withoutParam': f({ test: 'number.withoutParam' }),
  'number.withOneParam': f({ test: 'number.withOneParam' }, { test: { one: 1 } }),
  'number.withOneParam.thousand': f({ test: 'number.withOneParam' }, { test: { one: 1000 } }),
  'number.withPercentParam': f({ test: 'number.withPercentParam' }, { test: { one: 0.5 } }),
  'number.withUSDFormat': f({ test: 'number.withUSDFormat' }, { test: { one: 1234 } }),
  'select.gender.male': f({ test: 'select.gender' }, { test: { gender: 'male' } }),
  'select.gender.female': f({ test: 'select.gender' }, { test: { gender: 'female' } }),
  'select.gender.shemale': f({ test: 'select.gender' }, { test: { gender: 'shemale' } }),
})

const tester = (lang, locale) =>
  it(`should format message with ${lang} Locale`, () => {
    expect(keyValueResources(lang, format(lang, locale, customFormats))).toMatchSnapshot()
  })

describe('src/format', () => {
  addLocaleData(localDataFR)
  tester('EN', en)
  tester('FR', fr)
})
