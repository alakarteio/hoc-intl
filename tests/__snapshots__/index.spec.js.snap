// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`src/format should format message with EN Locale 1`] = `
Object {
  "language": Object {
    "locale": "en",
  },
  "number.withOneParam": Object {
    "test": "with 1 parameter",
  },
  "number.withOneParam.thousand": Object {
    "test": "with 1,000 parameter",
  },
  "number.withPercentParam": Object {
    "test": "with 50% parameter",
  },
  "number.withUSDFormat": Object {
    "test": "with $1,234.00 parameter",
  },
  "number.withoutParam": Object {
    "test": "no parameter",
  },
  "plural.simple.noparam": Object {
    "test": "You have {param, plural, =0 {no parameter.} =1 {one parameter.} other {# parameters.}}",
  },
  "plural.simple.one": Object {
    "test": "You have one parameter.",
  },
  "plural.simple.thousand": Object {
    "test": "You have 1,000 parameters.",
  },
  "plural.simple.two": Object {
    "test": "You have 2 parameters.",
  },
  "plural.simple.zero": Object {
    "test": "You have no parameter.",
  },
  "select.gender.female": Object {
    "test": "She will respond shortly.",
  },
  "select.gender.male": Object {
    "test": "He will respond shortly.",
  },
  "select.gender.shemale": Object {
    "test": "They will respond shortly.",
  },
  "simple.withOneParam": Object {
    "test": "with ONE parameter",
  },
  "simple.withTwoParam": Object {
    "test": "with ONE TWO parameter",
  },
  "simple.withoutParam": Object {
    "test": "no parameter",
  },
}
`;

exports[`src/format should format message with FR Locale 1`] = `
Object {
  "language": Object {
    "locale": "fr",
  },
  "number.withOneParam": Object {
    "test": "avec 1 paramètre",
  },
  "number.withOneParam.thousand": Object {
    "test": "avec 1,000 paramètre",
  },
  "number.withPercentParam": Object {
    "test": "avec 50% paramètre",
  },
  "number.withUSDFormat": Object {
    "test": "avec US$ 1,234.00 paramètre",
  },
  "number.withoutParam": Object {
    "test": "aucun paramètre",
  },
  "plural.simple.noparam": Object {
    "test": "vous avez {param, plural, =0 {aucun paramètre.} =1 {one paramètre.} other {# paramètres.}}",
  },
  "plural.simple.one": Object {
    "test": "vous avez one paramètre.",
  },
  "plural.simple.thousand": Object {
    "test": "vous avez 1,000 paramètres.",
  },
  "plural.simple.two": Object {
    "test": "vous avez 2 paramètres.",
  },
  "plural.simple.zero": Object {
    "test": "vous avez aucun paramètre.",
  },
  "select.gender.female": Object {
    "test": "Elle répondra sous peu.",
  },
  "select.gender.male": Object {
    "test": "Il répondra sous peu.",
  },
  "select.gender.shemale": Object {
    "test": "ça répondra sous peu.",
  },
  "simple.withOneParam": Object {
    "test": "avec ONE paramètre",
  },
  "simple.withTwoParam": Object {
    "test": "avec ONE TWO paramètre",
  },
  "simple.withoutParam": Object {
    "test": "aucun paramètre",
  },
}
`;
